File: SearchInA2dMatrix2.java
// Time Complexity : O(m+n) m - number of rows; n - numner of columns
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

/*
	1. Start at Top-Right corner
    2. if target < matrix element 
            move right
        else
            move down
	3. if matrix element = target, return true. If not found, return false
*/

// Your code here along with comments explaining your approach

public class SearchInA2dMatrix2 {

	public boolean searchMatrix(int[][] matrix, int target) {
        
        //edge
        if(matrix == null || matrix.length == 0)
            return false;
        
        //Bounds
        int m = matrix.length, n = matrix[0].length;
        
        //Starting at top-right corner
        int row = 0, col = n-1;
        
        //Conditions of while is based on the starting point
        while(row < m && col >= 0) {
            //Found the target
            if(target == matrix[row][col])
                return true;
            
            //Go left for smaller numbers
            else if(target < matrix[row][col])
                col--;
            
            //Go down for larger numbers
            else if(target > matrix[row][col])
                row++;
        }
}

File: search_in_a_2d_matrix_2.py
# Time Complexity:- O(m+n)
# Space Complexity:- O(1)
# Approach:- Start from the top right corner, the number at the index is less than target we might probably find the target down in the array but definitely not on the left as the entire column is sorted, so we increase i, if the number at the index is greater than the target we might probably find the target on the left but definitely not down in the column as the entire column is sorted. We keep on doing this until we are in the bounds.
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix or len(matrix[0])==0:
            return False
        i=0
        j=len(matrix[0])-1
        while(i<len(matrix) and j<len(matrix[0]) and i>=0 and j>=0):
            if matrix[i][j]==target:
                return True
            # number at index<target we need to go down
            if matrix[i][j]<target:
                i+=1
            else:
                # number at index>target we need to go left
                j-=1
        return False